### DISCUSSION TITLE
How to include non-Quarto-generated formats in "Other formats"?

### DISCUSSION THREAD

**Discussion by allefeld** [quarto-dev/quarto-cli#7411] [Q&A] [answered]
### Description

I'm using a Quarto website project to document a software library, including working code examples as `qmd` files, and I would like to provide on the `qmd`-derived pages a link to download the code as a script.

Quarto offers the possibility to add `ipynb` to `formats:`, which then creates an "Other formats" section below the table of contents, which includes a "Juypter" link. However, because the targeted community does not normally use Jupyter notebooks, offering to download an `ipynb` file would not be very useful for them. Instead, I would use a pre-render script to extract the code from the `qmd` file and write it to a regular script file of the language – but I don't see a way to link to such a script file under "Other formats".

I thought about writing a custom format, but those need to be based on a format built into Quarto, and none of them seem to be appropriate for the purpose. Script files are not a special case of documents, presentations, markdown files, wiki files, or any of the other formats.

I could of course simply link to the script file from the document text, but I'd prefer a more visible position like in the right margin.

I'd appreciate tips how to achieve something like this with the current Quarto functionality.

Otherwise I propose it as a new feature. That could either be options to customize "Other formats", or a new `script` output format. Such a `script` format would have to provide options to set the comment character (to include Markdown as comments) as well as the file extension.

**Comment by mcanouil** [comment, marked as answer]
Why not use code tools?

https://quarto.org/docs/output-formats/html-code.html#code-tools

See:
- https://github.com/quarto-dev/quarto-cli/discussions/3056

**Reply by allefeld** [reply]
I was first confused by your comment because the code tools are advertised as showing the `qmd` source, not a script file. But then I figured out that using `source:` its possible to link to an arbitrary URL, including a script file copied to the output as a resource. So you're right, that is a possibility, thanks!

What I dislike about that is that by default the URL seems to be loaded as a replacement for the current page, not as a download.  I would therefore downgrade my feature request to this: Please add a boolean `download:` option to `code-tools`, which if true adds the attribute `download` to the generated `a` element.

In the meantime I discovered that it is possible to put arbitrary material in the margin via `margin-header` and `margin-footer`, and though this is semantically somewhat wrong, it does the trick for me.

**Reply by mcanouil** [reply]
See:
- https://github.com/quarto-dev/quarto-cli/issues/7436
